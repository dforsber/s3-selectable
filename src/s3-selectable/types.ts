import {
  InputSerialization,
  OutputSerialization,
  S3,
  SelectObjectContentCommandInput,
  SelectObjectContentEventStream,
} from "@aws-sdk/client-s3";

import { Glue } from "@aws-sdk/client-glue";
import { LogLevel } from "bunyan";

export type PartialBy<TType, TKey extends keyof TType> = Omit<TType, TKey> & Partial<Pick<TType, TKey>>;

export type TS3SelectaParams = PartialBy<
  SelectObjectContentCommandInput,
  "Bucket" | "Key" | "ExpressionType" | "OutputSerialization" | "InputSerialization"
>;

export type TSelectParamsVerified = PartialBy<SelectObjectContentCommandInput, "Bucket" | "Key"> & {
  Expression: string;
  ExpressionType: string;
};

export type TS3electObjectContentVerified = Omit<ISelectObjectContent, "selectParams"> & {
  selectParams: TSelectParamsVerified;
};

export type TEvents = SelectObjectContentEventStream;

export const defaultS3SelectParms = {
  ExpressionType: "SQL",
  OutputSerialization: { JSON: {} },
};

export interface IS3Selectable {
  tableName: string;
  databaseName: string;
  glue: Glue;
  s3: S3;
  logLevel?: LogLevel;
}

export interface ISelectObjectContent {
  selectParams: TS3SelectaParams;
  onEventHandler?: (event: TEvents) => void;
  onDataHandler?: (event: Uint8Array) => void;
  onEndHandler?: () => void;
}

export interface IS3selectableNonClass {
  sql: string;
  s3: S3;
  glue: Glue;
  InputSerialization?: InputSerialization;
  OutputSerialization?: OutputSerialization;
  loglevel?: LogLevel;
}
