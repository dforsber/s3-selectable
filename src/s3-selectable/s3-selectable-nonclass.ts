import { InputSerialization, OutputSerialization, S3 } from "@aws-sdk/client-s3";

import { Glue } from "@aws-sdk/client-glue";
import { S3Selectable } from "./s3-selectable";
import { getTableAndDbAndExpr } from "../utils/sql-query.helper";

export interface IS3selectableNonClass {
  sql: string;
  s3: S3;
  glue: Glue;
  InputSerialization?: InputSerialization;
  OutputSerialization?: OutputSerialization;
  loglevel?: "trace" | "debug" | "info" | "warn" | "error" | "fatal"; // Match with Bunyan
}

/*
 * Alternative, non-class based interface that does not do any caching of Table data. We recommend
 * to use the class based interface for latency aware use cases where multiple S3 Select commands
 * are run over the same table and in cases where the class can be instantiated beforehand to fill
 * up the cache.
 */
export async function S3SelectableNonClass(params: IS3selectableNonClass): Promise<Uint8Array[]> {
  const { s3, glue } = params;
  const [databaseName, tableName, expr] = getTableAndDbAndExpr(params.sql);
  const Expression = params.sql.replace(`${databaseName}.${tableName}`, `s3Object${expr}`);
  const selectable = new S3Selectable({ databaseName, tableName, s3, glue });
  const data: Uint8Array[] = await new Promise(resolve => {
    const chunks: Uint8Array[] = [];
    selectable.select({
      selectParams: { ...params, Expression },
      onDataHandler: chunk => chunks.push(chunk),
      onEndHandler: () => resolve(chunks),
    });
  });
  return data;
}
