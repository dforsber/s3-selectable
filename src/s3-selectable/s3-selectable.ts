import {
  InputSerialization,
  OutputSerialization,
  S3,
  SelectObjectContentCommandInput,
  SelectObjectContentEventStream,
} from "@aws-sdk/client-s3";
import { getSQLWhereString, getTableAndDb } from "../utils/sql-query.helper";
import stream, { Readable } from "stream";

import { Glue } from "@aws-sdk/client-glue";
import { GlueTableToS3Key } from "../mappers/glueTableToS3Keys.mapper";
import { PartitionPreFilter } from "../utils/partition-filterer";
import mergeStream from "merge-stream";

export type PartialBy<TType, TKey extends keyof TType> = Omit<TType, TKey> & Partial<Pick<TType, TKey>>;

export type TS3SelectableParams = PartialBy<
  SelectObjectContentCommandInput,
  "Bucket" | "Key" | "ExpressionType" | "OutputSerialization"
>;

const defaults = {
  ExpressionType: "SQL",
  OutputSerialization: { JSON: {} },
};

export interface IS3Selectable {
  tableName: string;
  databaseName: string;
  glue: Glue;
  s3: S3;
}

export class S3Selectable {
  private s3 = this.params.s3;
  private glue = this.params.glue;
  private partitionValues!: string[];
  private partitionColumns!: string[];
  private s3bucket!: string;
  private partitionsFilter!: PartitionPreFilter;
  private mapper = new GlueTableToS3Key({
    s3: this.s3,
    glue: this.glue,
    databaseName: this.params.databaseName,
    tableName: this.params.tableName,
  });

  constructor(private params: IS3Selectable) {}

  public async selectObjectContent(
    params: TS3SelectableParams,
    onDataHandler?: (event: SelectObjectContentEventStream.RecordsMember) => void,
    onEndHandler?: (event: SelectObjectContentEventStream.RecordsMember) => void,
  ): Promise<stream> {
    await this.cacheTableMetadata();
    if (!params.Expression) throw new Error("S3 Select params Expression is required");
    const whereSql = getSQLWhereString(params.Expression, this.partitionColumns);
    const filteredPartitionValues = await this.partitionsFilter.filterPartitions(whereSql);
    const s3Keys = await this.mapper.getKeysByPartitions(filteredPartitionValues);
    const selectStreams = await Promise.all(
      s3Keys.map((Key: string) => this.getSelectStream({ ...defaults, ...params, Key }, onDataHandler, onEndHandler)),
    );
    return mergeStream(selectStreams);
  }

  private async getSelectStream(
    queryParams: PartialBy<SelectObjectContentCommandInput, "Bucket">,
    onDataHandler?: (event: SelectObjectContentEventStream.RecordsMember) => void,
    onEndHandler?: (event: SelectObjectContentEventStream.RecordsMember) => void,
  ): Promise<Readable> {
    const selStream = await this.s3.selectObjectContent({ ...queryParams, Bucket: this.s3bucket });
    if (selStream.Payload === undefined) throw new Error(`No select stream for ${queryParams.Key}`);
    const data = stream.Readable.from(selStream.Payload);
    if (onDataHandler) data.on("data", onDataHandler);
    if (onEndHandler) data.on("end", onEndHandler);
    return data;
  }

  /*
   * Increased complexity is due to fetching both getTableInfo and
   * getPartitionValues concurrently (Promise.all) while doing caching
   */
  public async cacheTableMetadata(): Promise<void> {
    const partCols = await this.mapper.getTableInfo();
    [this.s3bucket, this.partitionColumns] = [partCols.Bucket, partCols.PartitionColumns];
    this.partitionValues = await this.mapper.getPartitionValues();
    this.partitionsFilter = this.partitionsFilter
      ? this.partitionsFilter
      : new PartitionPreFilter(this.partitionValues, this.partitionColumns);
  }
}

/*
 * Alternative, non-class based interface that does not do any caching of Table data. We recommend
 * to use the class based interface for latency aware use cases where multiple S3 Select commands
 * are run over the same table and in cases where the class can be instantiated beforehand to fill
 * up the cache.
 */
export async function s3selectableNonClass(
  sql: string,
  s3: S3,
  glue: Glue,
  inpSer: InputSerialization = { CSV: {}, CompressionType: "GZIP" },
  outSer: OutputSerialization = { JSON: {} },
): Promise<string[]> {
  const [databaseName, tableName] = getTableAndDb(sql);
  const Expression = sql.replace(`${databaseName}.${tableName}`, "s3Object");
  const selectable = new S3Selectable({ databaseName, tableName, s3, glue });
  const rowsStream = await selectable.selectObjectContent({
    Expression,
    ExpressionType: "SQL",
    InputSerialization: inpSer,
    OutputSerialization: outSer,
  });
  const data: string[] = await new Promise(resolve => {
    const rows: string[] = [];
    rowsStream.on("data", chunk => {
      if (chunk.Records?.Payload) rows.push(Buffer.from(chunk.Records.Payload).toString());
    });
    rowsStream.on("end", () => resolve(rows));
  });
  return data;
}
