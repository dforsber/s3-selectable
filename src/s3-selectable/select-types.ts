import { SelectObjectContentCommandInput, SelectObjectContentEventStream } from "@aws-sdk/client-s3";

import { ITableInfo } from "../mappers/glueTableToS3Keys.mapper";

/*
 * interface for select() method
 */
export interface ISelect {
  selectParams: TS3SelectObjectContent;
  onEventHandler?: (event: TEvents) => void;
  onDataHandler?: (event: Uint8Array) => void;
  onEndHandler?: () => void;
}

export type TEvents = SelectObjectContentEventStream;

export const defaultS3SelectParms = {
  ExpressionType: "SQL",
  OutputSerialization: { JSON: {} },
};

export type TS3SelectObjectContent = PartialBy<
  SelectObjectContentCommandInput,
  "Bucket" | "Key" | "ExpressionType" | "OutputSerialization" | "InputSerialization"
>;

export type PartialBy<TType, TKey extends keyof TType> = Omit<TType, TKey> & Partial<Pick<TType, TKey>>;

export type TS3SelectObjectContentVerified = PartialBy<SelectObjectContentCommandInput, "Key"> & {
  Expression: string;
  ExpressionType: string;
  Bucket: string;
};

export type TS3electObjectContentVerified = Omit<ISelect, "selectParams"> & {
  selectParams: TS3SelectObjectContentVerified;
};

export interface IPreparedSelect extends TS3electObjectContentVerified {
  limit: number;
  s3Keys: string[];
}

export interface IExplainSelect {
  tableInfo: ITableInfo;
  preparedSelect: IPreparedSelect;
  partitionFilter: string;
}
