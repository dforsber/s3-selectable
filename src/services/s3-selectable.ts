import AWS from "aws-sdk";
import mergeStream from "merge-stream";
import { filterPartitions } from "../utils/sqlite.filter";
import { getSQLWhereString } from "../utils/sql-query.helper";
import { getTableAndDb } from "../utils/parse-db-and-table";
import { GlueTableToS3Key } from "../mappers/glueTableToS3Keys.mapper";
import { StreamingEventStream } from "aws-sdk/lib/event-stream/event-stream";
import {
  ContinuationEvent,
  EndEvent,
  Expression,
  InputSerialization,
  OutputSerialization,
  ProgressEvent,
  RecordsEvent,
  SelectObjectContentRequest,
  StatsEvent,
} from "aws-sdk/clients/s3";

export type PartialBy<TType, TKey extends keyof TType> = Omit<TType, TKey> & Partial<Pick<TType, TKey>>;

export interface IEventStream {
  Records?: RecordsEvent;
  Stats?: StatsEvent;
  Progress?: ProgressEvent;
  Cont?: ContinuationEvent;
  End?: EndEvent;
}

export type SelectStream = StreamingEventStream<IEventStream>;

export interface IS3Selectable {
  tableName: string;
  databaseName: string;
  glue: AWS.Glue;
  s3: AWS.S3;
}

export class S3Selectable {
  private s3 = this.params.s3;
  private glue = this.params.glue;
  private partitionValues!: string[];
  private partitionColumns!: string[];
  private s3bucket!: string;
  private mapper = new GlueTableToS3Key({
    s3: this.s3,
    glue: this.glue,
    databaseName: this.params.databaseName,
    tableName: this.params.tableName,
  });

  constructor(private params: IS3Selectable) {}

  public async selectObjectContent(
    params: PartialBy<SelectObjectContentRequest, "Bucket" | "Key">,
  ): Promise<SelectStream> {
    await this.getData();
    const filteredPartitionValues = await this.filterPartitionsByQuery(params.Expression);
    const s3Keys = await this.mapper.getKeysByPartitions(filteredPartitionValues);
    const selectStreams = await Promise.all(s3Keys.map((Key: string) => this.getSelectStream({ ...params, Key })));
    return mergeStream(...selectStreams.filter(stream => !!stream));
  }

  private async getSelectStream(queryParams: PartialBy<SelectObjectContentRequest, "Bucket">): Promise<SelectStream> {
    const stream = await this.s3.selectObjectContent({ ...queryParams, Bucket: this.s3bucket }).promise();
    if (stream.Payload === undefined) throw new Error(`No select stream for ${queryParams.Key}`);
    return <SelectStream>stream.Payload;
  }

  /*
   * Increased complexity is due to fetching both getTableInfo and
   * getPartitionValues concurrently (Promise.all) while doing caching
   */
  private async getData(): Promise<void> {
    const partColsFetched = this.partitionColumns && this.s3bucket;
    const partColsProm = partColsFetched
      ? Promise.resolve({ Bucket: this.s3bucket, PartitionColumns: this.partitionColumns })
      : this.mapper.getTableInfo();
    const partVolsProm = this.partitionValues
      ? Promise.resolve(this.partitionValues)
      : this.mapper.getPartitionValues();
    let partCols;
    [partCols, this.partitionValues] = await Promise.all([partColsProm, partVolsProm]);
    [this.s3bucket, this.partitionColumns] = [partCols.Bucket, partCols.PartitionColumns];
  }

  public async filterPartitionsByQuery(expression: Expression): Promise<string[]> {
    await this.getData();
    if (!this.partitionValues) return Promise.resolve(this.partitionValues);
    return filterPartitions(
      this.partitionValues,
      this.partitionColumns,
      getSQLWhereString(expression, this.partitionColumns),
    );
  }
}

/*
 * Alternative, non-class based interface that does not do any caching of Table data. We recommend
 * to use the class based interface for latency aware use cases where multiple S3 Select commands
 * are run over the same table and in cases where the class can be instantiated beforehand to fill
 * up the cache.
 */
export async function s3selectable(
  sql: string,
  inpSer: InputSerialization = { CSV: {}, CompressionType: "GZIP" },
  outSer: OutputSerialization = { JSON: {} },
): Promise<string[]> {
  const [databaseName, tableName] = getTableAndDb(sql);
  const Expression = sql.replace(`${databaseName}.${tableName}`, "s3Object");
  const selectable = new S3Selectable({ databaseName, tableName, s3: new AWS.S3(), glue: new AWS.Glue() });
  const rowsStream = await selectable.selectObjectContent({
    Expression,
    ExpressionType: "SQL",
    InputSerialization: inpSer,
    OutputSerialization: outSer,
  });
  const data: string[] = await new Promise(resolve => {
    const d: string[] = [];
    rowsStream.on("end", () => resolve(d));
    rowsStream.on("data", chunk => d.push(Buffer.from(chunk).toString()));
  });
  return data;
}
