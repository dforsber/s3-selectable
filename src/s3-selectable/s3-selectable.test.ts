import * as AWSMock from "aws-sdk-mock";
import { GetPartitionsRequest, GetTableRequest } from "aws-sdk/clients/glue";
import { Readable, ReadableOptions } from "stream";
import { S3Selectable, s3selectable } from "./s3-selectable";
import { testTable, testTableKeys, testTablePartitions } from "../common/fixtures/glue-table";
/* eslint-disable @typescript-eslint/ban-types */
import {
  InputSerialization,
  ListObjectsV2Request,
  OutputSerialization,
  SelectObjectContentRequest,
} from "aws-sdk/clients/s3";

/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
const AWS = require("aws-sdk");

class MockedSelectStream extends Readable {
  public Payload: Readable | undefined = this;

  private rows = [
    { id: 1, value: "test1" },
    { id: 2, value: "test2" },
  ];
  private index: number;

  constructor(opt?: ReadableOptions) {
    super(opt);
    this.index = 0;
  }

  public _read(_size: number): void {
    const i = this.index++;
    if (i >= this.rows.length) {
      this.push(null);
      return;
    }
    const buf = Buffer.from(JSON.stringify(this.rows[i]), "ascii");
    this.push(buf);
  }
}

class MockedSelectStreamNoPayload extends MockedSelectStream {
  public Payload = undefined;
  constructor(opt?: ReadableOptions) {
    super(opt);
  }
}

let glueGetTableCalled = 0;
let glueGetPartitionsCalled = 0;
let s3ListObjectsV2Called = 0;
let selectObjectContent = 0;
AWSMock.setSDKInstance(AWS);
AWSMock.mock("Glue", "getTable", (params: GetTableRequest, cb: Function) => {
  glueGetTableCalled++;
  if (params.Name !== "partitioned_and_bucketed_elb_logs_parquet")
    cb(new Error(`Table not found: ${params.Name}`), null);
  cb(null, { Table: testTable });
});
AWSMock.mock("Glue", "getPartitions", (_params: GetPartitionsRequest, cb: Function) => {
  glueGetPartitionsCalled++;
  cb(null, { Partitions: testTablePartitions });
});
AWSMock.mock("S3", "listObjectsV2", (params: ListObjectsV2Request, cb: Function) => {
  s3ListObjectsV2Called++;
  const pref = params.Prefix ?? "";
  cb(null, {
    NextContinuationToken: undefined,
    Contents: testTableKeys.filter(k => k.includes(pref)).map(k => ({ Key: k })),
  });
});
AWSMock.mock("S3", "selectObjectContent", (_params: SelectObjectContentRequest, cb: Function) => {
  selectObjectContent++;
  cb(null, new MockedSelectStream());
});

const s3 = new AWS.S3({ region: "eu-west-1" });
const glue = new AWS.Glue({ region: "eu-west-1" });
const [databaseName, tableName] = ["default", "partitioned_and_bucketed_elb_logs_parquet"];
const params = { glue, s3, tableName, databaseName };
beforeEach(() => {
  glueGetTableCalled = 0;
  glueGetPartitionsCalled = 0;
  s3ListObjectsV2Called = 0;
  selectObjectContent = 0;
});

describe("Test selectObjectContent", () => {
  it("first verify mockedReadable", async () => {
    const readable = new MockedSelectStream();
    const rows = await new Promise(r => {
      const rows: string[] = [];
      readable.on("data", chunk => rows.push(Buffer.from(chunk).toString()));
      readable.on("end", () => r(rows));
    });
    await new Promise(r => setTimeout(r, 1000));
    expect(rows).toMatchInlineSnapshot(`
      Array [
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
      ]
    `);
  });

  it("selectObjectContent provides correct results", async () => {
    const sql = "SELECT * FROM s3Object WHERE elb_response_code='302' AND ssl_protocol='-'";
    const inpSer: InputSerialization = { CSV: {}, CompressionType: "GZIP" };
    const outSer: OutputSerialization = { JSON: {} };
    const selectable = new S3Selectable(params);
    await selectable.selectObjectContent({
      Expression: sql,
      ExpressionType: "SQL",
      InputSerialization: inpSer,
      OutputSerialization: outSer,
    });
    expect(glueGetTableCalled).toEqual(1);
    expect(glueGetPartitionsCalled).toEqual(1); // 1 table
    expect(s3ListObjectsV2Called).toEqual(1); // 1 partition
    expect(selectObjectContent).toEqual(10); // 10 objects
    const rowsStream = await selectable.selectObjectContent({
      Expression: sql,
      ExpressionType: "SQL",
      InputSerialization: inpSer,
      OutputSerialization: outSer,
    });
    expect(glueGetTableCalled).toEqual(1);
    expect(glueGetPartitionsCalled).toEqual(1); // 1 table
    expect(s3ListObjectsV2Called).toEqual(1); // S3 Keys are cached per partition
    expect(selectObjectContent).toEqual(20); // 2 * 10 objects
    const rows = await new Promise(r => {
      const rows: string[] = [];
      rowsStream.on("data", chunk => rows.push(Buffer.from(chunk).toString()));
      rowsStream.on("end", () => r(rows));
    });
    expect(rows).toMatchInlineSnapshot(`
      Array [
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
      ]
    `);
  });

  it("throws when stream does not contain Payload", async () => {
    AWSMock.restore("S3", "selectObjectContent");
    AWSMock.mock("S3", "selectObjectContent", (_params: SelectObjectContentRequest, cb: Function) => {
      selectObjectContent++;
      cb(null, new MockedSelectStreamNoPayload());
    });
    const s3 = new AWS.S3({ region: "eu-west-1" });
    const glue = new AWS.Glue({ region: "eu-west-1" });
    const [databaseName, tableName] = ["default", "partitioned_and_bucketed_elb_logs_parquet"];
    const params = { glue, s3, tableName, databaseName };
    const sql = "SELECT * FROM s3Object WHERE elb_response_code='302' AND ssl_protocol='-'";
    const inpSer: InputSerialization = { CSV: {}, CompressionType: "GZIP" };
    const outSer: OutputSerialization = { JSON: {} };
    const selectable = new S3Selectable(params);
    await expect(
      async () =>
        await selectable.selectObjectContent({
          Expression: sql,
          ExpressionType: "SQL",
          InputSerialization: inpSer,
          OutputSerialization: outSer,
        }),
    ).rejects.toThrowError();
    expect(glueGetTableCalled).toEqual(1);
    expect(glueGetPartitionsCalled).toEqual(1); // 1 table
    expect(s3ListObjectsV2Called).toEqual(1); // 1 partition
    expect(selectObjectContent).toEqual(10); // 10 objects
    AWSMock.restore("S3", "selectObjectContent");
    AWSMock.mock("S3", "selectObjectContent", (_params: SelectObjectContentRequest, cb: Function) => {
      selectObjectContent++;
      cb(null, new MockedSelectStream());
    });
  });
});

describe("Non-class based s3selectable returns correct results", () => {
  it("test", async () => {
    const table = "default.partitioned_and_bucketed_elb_logs_parquet";
    const sql = `SELECT * FROM ${table} WHERE elb_response_code='302' AND ssl_protocol='-'`;
    const rows = await s3selectable(sql);
    expect(rows).toMatchInlineSnapshot(`
      Array [
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
        "{\\"id\\":1,\\"value\\":\\"test1\\"}",
        "{\\"id\\":2,\\"value\\":\\"test2\\"}",
      ]
    `);
  });
});
