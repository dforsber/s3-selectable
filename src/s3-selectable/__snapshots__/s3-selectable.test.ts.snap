// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test selectObjectContent explain select works 1`] = `
{
  "partitionFilter": "SELECT partition FROM partitions WHERE \`elb_response_code\` = '302' AND \`ssl_protocol\` = '-'",
  "preparedSelect": {
    "limit": 0,
    "s3Keys": [
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00000",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00001",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00002",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00003",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00004",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00005",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00006",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00007",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00008",
      "Unsaved/2020/05/25/tables/63e1dd93-76d5-497f-8db7-bab5861fe14e/ssl_protocol=-/elb_response_code=302/20200525_195025_00011_xgnnv_bucket-00009",
    ],
    "selectParams": {
      "Bucket": "dummy-test-bucket",
      "Expression": "SELECT * FROM db.t WHERE TRUE AND TRUE",
      "ExpressionType": "SQL",
      "InputSerialization": {
        "Parquet": {},
      },
      "OutputSerialization": {
        "JSON": {},
      },
    },
  },
  "tableInfo": {
    "Bucket": "dummy-test-bucket",
    "InputSerialization": {
      "Parquet": {},
    },
    "PartitionColumns": [
      "ssl_protocol",
      "elb_response_code",
    ],
  },
}
`;

exports[`Test selectObjectContent explain select works in case of filtering does not produce any keys 1`] = `
{
  "partitionFilter": "SELECT partition FROM partitions WHERE \`elb_response_code\` = 'nonsense' AND \`ssl_protocol\` = 'badInput'",
  "preparedSelect": {
    "limit": 0,
    "s3Keys": [],
    "selectParams": {
      "Bucket": "dummy-test-bucket",
      "Expression": "SELECT * FROM db.t WHERE TRUE AND TRUE",
      "ExpressionType": "SQL",
      "InputSerialization": {
        "Parquet": {},
      },
      "OutputSerialization": {
        "JSON": {},
      },
    },
  },
  "tableInfo": {
    "Bucket": "dummy-test-bucket",
    "InputSerialization": {
      "Parquet": {},
    },
    "PartitionColumns": [
      "ssl_protocol",
      "elb_response_code",
    ],
  },
}
`;

exports[`Test selectObjectContent first verify mockedReadable 1`] = `
[
  "{"id":1,"value":"test1"}",
  "{"id":2,"value":"test2"}",
]
`;

exports[`Test selectObjectContent selectObjectContent provides correct results 1`] = `
[
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
]
`;

exports[`Test selectObjectContent selectObjectContent provides correct results with onDataHandler and onEndHandler 1`] = `
[
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
]
`;

exports[`Test selectObjectContent selectObjectContent provides correct results with onDataHandler and onEndHandler 2`] = `
[
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
]
`;

exports[`Test selectObjectContent selectObjectContent uses LIMIT 2 by returning only 2 rows 1`] = `
[
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
]
`;

exports[`Test selectObjectContent selectObjectContent uses LIMIT 42 by returning all rows 1`] = `
[
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":1,"value":"test1"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
  "{"id":2,"value":"test2"}",
]
`;
